aten/src/ATen/native/cuda/Unique.cu

```diff
@@ -17,12 +22,100 @@ namespace native{
 
 namespace {
 
+// unique
+template <typename scalar_t>
+std::tuple<Tensor, Tensor, Tensor> unique_cuda_template(
+  const Tensor& self,
+  const bool consecutive,
+  const bool return_inverse,
+  const bool return_counts
+) {
+
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+
+  void *tmp_storage = nullptr;
+  size_t tmp_storage_size = 0;
+
+  int64_t num_inp = self.numel();
+  const scalar_t* self_data = self.data_ptr<scalar_t>();
+  int64_t *sorted_indices_data = nullptr;
+  Tensor sorted;
+  const scalar_t* sorted_data = self_data;
+
+  Tensor inverse_indices;
+
+  if (return_inverse) {
+    inverse_indices = at::empty(self.sizes(), self.options().dtype(kLong));
+  }
+
+  if (!consecutive) {
+    sorted = at::empty({num_inp}, self.options());
+    scalar_t *sorted_data_ = sorted.data_ptr<scalar_t>();
+    sorted_data = sorted_data_;  // const pointer
+    if (return_inverse) {
+      auto arange = at::arange(0, num_inp, self.options().dtype(kLong));
+      int64_t *arange_data = arange.data_ptr<int64_t>();
+      sorted_indices_data = static_cast<int64_t *>(c10::cuda::CUDACachingAllocator::raw_alloc(num_inp * sizeof(int64_t)));
+      // get tmp_storage_size
+      cub::DeviceRadixSort::SortPairs(tmp_storage, tmp_storage_size, self_data, sorted_data_, arange_data, sorted_indices_data, num_inp, 0, sizeof(scalar_t) * 8, stream);
+      // allocate tmp_storage and run the computation
+      tmp_storage = c10::cuda::CUDACachingAllocator::raw_alloc(tmp_storage_size);
```

Comment:
How much using raw_alloc is saving you compared to at::native::empty_cuda? I would really like to avoid manual memory management, it makes code unmaintainable.

Reply:
It would add `~3-5us` each run. I think this is affordable, 5/300 is nothing

Reply:
I have updated this PR to use Tensor for memory management.